### PR TITLE
[new-parser] Fix AccumulateTest and add groupBy support

### LINKED ISSUE #5803: [new-parser] Support groupby
## Parent issue
- #5678 

## Partially blocked by
- #5723

## Failing tests
- `org.drools.model.codegen.execmodel.GroupByTest`

## Notes
- `STANDARD_FROM_DRL` tests are blocked by #5723.
- `PATTERN_DSL` tests fail because `groupBy` is not supported.

## Rule code snippet
```
groupby($i: Integer() and exists String();
$key : Math.abs($i);
$count : count())
then
```

## Error output
```
### parse : ANTLR4_PARSER_ENABLED = true
line 4:76 missing ')' at ';'
line 4:105 mismatched input ')' expecting {'unit', 'function', 'global', 'declare', 'rule', 'query', 'when', 'then', 'end', 'and', 'or', 'exists', 'not', 'in', 'from', 'accumulate', 'acc', 'init', 'action', 'reverse', 'result', 'entry-point', 'eval', 'forall', 'over', 'matches', 'memberOf', 'contains', 'excludes', 'soundslike', 'str', 'salience', 'enabled', 'no-loop', 'auto-focus', 'lock-on-active', 'refract', 'direct', 'agenda-group', 'activation-group', 'ruleflow-group', 'date-effective', 'date-expires', 'dialect', 'calendars', 'timer', 'duration', 'module', 'open', 'requires', 'exports', 'opens', 'to', 'uses', 'provides', 'with', 'transitive', 'var', 'yield', 'record', 'sealed', 'permits', '(', ';', '?', '@', IDENTIFIER}
11:24:03.797 [main] ERROR o.d.d.parser.antlr4.DRLParserWrapper.parse:77 - Exception while creating PackageDescr
org.drools.drl.parser.antlr4.DRLParserException: rhs has to start with 'then' : rhs = )
    $remappedKey: Object() from $key
    $remappedCount: Long() from $count
then
    typeChecker.accept($remappedKey);
	at org.drools.drl.parser.antlr4.Antlr4ParserStringUtils.trimThen(Antlr4ParserStringUtils.java:75)
	at org.drools.drl.parser.antlr4.DRLVisitorImpl.visitRuledef(DRLVisitorImpl.java:277)
	at org.drools.drl.parser.antlr4.DRLVisitorImpl.visitRuledef(DRLVisitorImpl.java:76)
	at org.drools.drl.parser.antlr4.DRLParser$RuledefContext.accept(DRLParser.java:1230)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.drools.drl.parser.antlr4.DRLParserBaseVisitor.visitDrlStatementdef(DRLParserBaseVisitor.java:27)
	at org.drools.drl.parser.antlr4.DRLParser$DrlStatementdefContext.accept(DRLParser.java:414)
	at org.drools.drl.parser.antlr4.DRLVisitorImpl.visitDescrChildren(DRLVisitorImpl.java:855)
	at org.drools.drl.parser.antlr4.DRLVisitorImpl.visitCompilationUnit(DRLVisitorImpl.java:93)
	at org.drools.drl.parser.antlr4.DRLVisitorImpl.visitCompilationUnit(DRLVisitorImpl.java:76)
	at org.drools.drl.parser.antlr4.DRLParser$CompilationUnitContext.accept(DRLParser.java:327)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.drools.drl.parser.antlr4.DRLParserHelper.compilationUnitContext2PackageDescr(DRLParserHelper.java:87)
	at org.drools.drl.parser.antlr4.DRLParserWrapper.parse(DRLParserWrapper.java:75)
	at org.drools.drl.parser.antlr4.DRLParserWrapper.parse(DRLParserWrapper.java:55)
	at org.drools.drl.parser.DrlParser.lambda$parse$2(DrlParser.java:189)
	at org.drools.drl.parser.DrlParser.compileWithAntlr4Parser(DrlParser.java:202)
	at org.drools.drl.parser.DrlParser.parse(DrlParser.java:189)
	at org.drools.drl.parser.DrlParser.parse(DrlParser.java:177)
	at org.drools.drl.parser.DrlParser.parse(DrlParser.java:166)
	at org.drools.compiler.builder.impl.resources.DrlResourceHandler.process(DrlResourceHandler.java:55)
	at org.drools.compiler.builder.impl.PackageDescrBuilder.buildResource(PackageDescrBuilder.java:73)
	at org.drools.compiler.builder.impl.PackageDescrBuilder.build(PackageDescrBuilder.java:51)
	at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.buildPackageDescr(CompositeKnowledgeBuilderImpl.java:173)
	at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:116)
	at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:109)
	at org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:274)
	at org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:220)
	at org.drools.compiler.kie.builder.impl.AbstractKieProject.verify(AbstractKieProject.java:84)
	at org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildKieProject(KieBuilderImpl.java:285)
	at org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:251)
	at org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:208)
	at org.drools.model.codegen.execmodel.BaseModelTest.createKieBuilder(BaseModelTest.java:149)
	at org.drools.model.codegen.execmodel.BaseModelTest.createKieBuilder(BaseModelTest.java:132)
	at org.drools.model.codegen.execmodel.BaseModelTest.getKieContainer(BaseModelTest.java:121)
	at org.drools.model.codegen.execmodel.BaseModelTest.getKieContainer(BaseModelTest.java:114)
	at org.drools.model.codegen.execmodel.BaseModelTest.getKieSession(BaseModelTest.java:110)
	at org.drools.model.codegen.execmodel.BaseModelTest.getKieSession(BaseModelTest.java:106)
	at org.drools.model.codegen.execmodel.GroupByTest.assertSessionHasProperties(GroupByTest.java:58)
	at org.drools.model.codegen.execmodel.GroupByTest.testBindingRemappedAfterGroupBy(GroupByTest.java:791)
11:24:03.832 [main] ERROR o.d.c.k.b.impl.AbstractKieProject.buildKnowledgePackages:280 - Unable to build KieBaseModel:defaultKieBase
[4,76]: missing ')' at ';'
[4,105]: mismatched input ')' expecting {'unit', 'function', 'global', 'declare', 'rule', 'query', 'when', 'then', 'end', 'and', 'or', 'exists', 'not', 'in', 'from', 'accumulate', 'acc', 'init', 'action', 'reverse', 'result', 'entry-point', 'eval', 'forall', 'over', 'matches', 'memberOf', 'contains', 'excludes', 'soundslike', 'str', 'salience', 'enabled', 'no-loop', 'auto-focus', 'lock-on-active', 'refract', 'direct', 'agenda-group', 'activation-group', 'ruleflow-group', 'date-effective', 'date-expires', 'dialect', 'calendars', 'timer', 'duration', 'module', 'open', 'requires', 'exports', 'opens', 'to', 'uses', 'provides', 'with', 'transitive', 'var', 'yield', 'record', 'sealed', 'permits', '(', ';', '?', '@', IDENTIFIER}
[0,0]: 
[0,0]: Parser returned a null Package


java.lang.AssertionError: [Message [id=1, kieBase=defaultKieBase, level=ERROR, path=r0.drl, line=4, column=0
   text=missing ')' at ';'], Message [id=2, kieBase=defaultKieBase, level=ERROR, path=r0.drl, line=4, column=0
   text=mismatched input ')' expecting {'unit', 'function', 'global', 'declare', 'rule', 'query', 'when', 'then', 'end', 'and', 'or', 'exists', 'not', 'in', 'from', 'accumulate', 'acc', 'init', 'action', 'reverse', 'result', 'entry-point', 'eval', 'forall', 'over', 'matches', 'memberOf', 'contains', 'excludes', 'soundslike', 'str', 'salience', 'enabled', 'no-loop', 'auto-focus', 'lock-on-active', 'refract', 'direct', 'agenda-group', 'activation-group', 'ruleflow-group', 'date-effective', 'date-expires', 'dialect', 'calendars', 'timer', 'duration', 'module', 'open', 'requires', 'exports', 'opens', 'to', 'uses', 'provides', 'with', 'transitive', 'var', 'yield', 'record', 'sealed', 'permits', '(', ';', '?', '@', IDENTIFIER}], Message [id=3, kieBase=defaultKieBase, level=ERROR, path=r0.drl, line=0, column=0
   text=], Message [id=4, kieBase=defaultKieBase, level=ERROR, path=r0.drl, line=0, column=0
   text=Parser returned a null Package]]

	at org.drools.model.codegen.execmodel.BaseModelTest.createKieBuilder(BaseModelTest.java:157)
	at org.drools.model.codegen.execmodel.BaseModelTest.createKieBuilder(BaseModelTest.java:132)
	at org.drools.model.codegen.execmodel.BaseModelTest.getKieContainer(BaseModelTest.java:121)
	at org.drools.model.codegen.execmodel.BaseModelTest.getKieContainer(BaseModelTest.java:114)
	at org.drools.model.codegen.execmodel.BaseModelTest.getKieSession(BaseModelTest.java:110)
	at org.drools.model.codegen.execmodel.BaseModelTest.getKieSession(BaseModelTest.java:106)
	at org.drools.model.codegen.execmodel.GroupByTest.assertSessionHasProperties(GroupByTest.java:58)
	at org.drools.model.codegen.execmodel.GroupByTest.testBindingRemappedAfterGroupBy(GroupByTest.java:791)
```

### LINKED ISSUE #5872: [new-parser] Broken AccumulateTest
## Parent issue
- #5678 

## Failing tests
- `org.drools.model.codegen.execmodel.AccumulateTest`

## Notes
Three issues:
- Accumulate function not recognized if it has no parameters, e.g. `count()`.
- Init - Action - Reverse - Result form requires `COMMA` as a separator but it should be optional.
- Semicolon after statements inside `init()` etc. should be optional.

### PR DESCRIPTION
**Pull request by yurloc**
- Fixes #5872. 
- Fixes #5803.

### Before PR in `drools-model-codegen`
```
[ERROR] Tests run: 2434, Failures: 74, Errors: 0, Skipped: 9
```

### After PR in `drools-model-codegen`
```
[ERROR] Tests run: 2434, Failures: 2, Errors: 0, Skipped: 9
```

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request and downstream checks</b>  
  - Push a new commit to the PR. An empty commit would be enough.

- for a <b>full downstream build</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- for <b>Jenkins PR check only</b>
  - If you are an ASF committer for KIE podling, login to Jenkins (https://ci-builds.apache.org/job/KIE/job/drools/), go to the specific PR job, and click on `Build Now` button.
</details>

### CODE REVIEW

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4 [hunk 1]
```diff
@@ -409,6 +413,26 @@ lhsAccumulate : (DRL_ACCUMULATE|DRL_ACC) LPAREN lhsAndDef (COMMA|SEMI)
                  RPAREN (SEMI)?
                  ;
 
+/**
+ * lhsGroupBy := GROUPBY LEFT_PAREN lhsAnd (COMMA|SEMICOLON)
+ *                      groupByKeyBinding SEMICOLON
+ *                      accumulateFunctionBinding (COMMA accumulateFunctionBinding)*
+ *                      (SEMICOLON constraints)?
+ *                  RIGHT_PAREN SEMICOLON?
+ */
```
Comment (mariofusco) [inline drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4]: If this is no longer necessary please remove it.

**Review by mariofusco — APPROVED** [review]

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4 (hunk 1 above)
Comment (yurloc) [reply drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4]: Sure, this no longer makes sense. I've included it for consistency with how the new parser has been developed but these rule comments from the old parser are either outdated and therefore wrong or they are correct in which case the new rule is often identical, making them redundant.

I suppose we're going to remove the other ones as well as part of the post-development cleanup.

**Comment by yurloc** [comment]
@tkobayas @gitgabrio please review.

File: drools-drl/drools-drl-parser-tests/src/test/resources/org/drools/drl/parser/antlr4/groupBy.drl [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+rule "GroupByParserTest"
+when
+    groupby( $p: Person ( $age : age < 30 );           // SOURCE_PATTERN
+             $initial : $p.getName().substring(0, 1);  // GROUPING_FUNCTION
+             $sumOfAges : sum($age),                   // 2 ACC_FUNCTIONS, one summing the ages of the persons in the group
+             $countOfPersons : count();                // and the other simply counting them
+             $sumOfAges > 10 )                         // CONSTRAINT filtering away groups
+then
+end
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/resources/org/drools/drl/parser/antlr4/groupBy.drl:26]: Copied from `drools-test-coverage`.

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java [hunk 3]
```diff
@@ -2085,6 +2086,43 @@ public void parse_EndPosition() throws Exception {
         assertThat(col.getEndLine()).isEqualTo(25);
     }
 
+    @Test
+    public void parse_GroupBy() throws Exception {
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java:2090]: Copied from `org.drools.mvel.compiler.lang.RuleParserTest#testGroupBy`.

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4 (hunk 1 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4]: Yeah, the comment from old parser would be useful when we hit an issue and check the first clue of how old parser works. In reality, we often need to debug DRL6Parser eventually, though.

**Review by tkobayas — APPROVED** [review]
Thank you, big improvement!

**Pull request merged**